### PR TITLE
[Merged by Bors] - feat(SimpleGraph/Coloring): Surjectivity of graph colorings

### PR DESCRIPTION
**Pull request by jt496**
Add two simple results on graph colorings with χ(G) colors.
1. If r ≤ χ(G) and C is an r-coloring of G then C is surjective
2. If G has an α-coloring then χ(G) = |α| iff every such coloring is surjective.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Coloring.lean [hunk 1]
```diff
@@ -392,6 +392,42 @@ theorem chromaticNumber_eq_card_of_forall_surj [Fintype α] (C : G.Coloring α)
     exact Nat.lt_le_asymm hc h2
 #align simple_graph.chromatic_number_eq_card_of_forall_surj SimpleGraph.chromaticNumber_eq_card_of_forall_surj
 
+
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Coloring.lean:395]:
```suggestion
```

File: Mathlib/Combinatorics/SimpleGraph/Coloring.lean [hunk 2]
```diff
@@ -392,6 +392,42 @@ theorem chromaticNumber_eq_card_of_forall_surj [Fintype α] (C : G.Coloring α)
...
+/-- If r ≤ χ(G) and C is an r-coloring of G then C is surjective -/
+lemma le_chromaticNumber_surj {r : ℕ} (C : G.Coloring (Fin r)) (hc : r ≤ G.chromaticNumber) :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Coloring.lean:397]:
```suggestion
lemma surjective_of_le_chromaticNumber (C : G.Coloring ι) (hc : card ι ≤ G.chromaticNumber) :
```
and `open Fintype` and pull the `{r : ℕ}` to the nearest `variable`

File: Mathlib/Combinatorics/SimpleGraph/Coloring.lean [hunk 3]
```diff
@@ -392,6 +392,42 @@ theorem chromaticNumber_eq_card_of_forall_surj [Fintype α] (C : G.Coloring α)
...
+    all_goals intro hne; simp only [Fin.mk.injEq, Fin.val_eq_val, hc] at hne
+    · exact hc b hne.symm
+    · exact C.valid had hne⟩
+    apply lt_of_le_of_lt (Colorable.chromaticNumber_le ⟨C'⟩)
+    rw [Nat.cast_lt]
+    exact Nat.lt_succ_self _
+
+/-- If G has a α-Coloring then χ(G) = ‖α‖ ↔ every α-Coloring of G is surjective -/
+theorem chromaticNumber_eq_card_iff_forall_surj [Fintype α] (C : G.Coloring α) :
+    G.chromaticNumber = Fintype.card α  ↔ (∀ C' : G.Coloring α, Function.Surjective C') := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Coloring.lean:420]:
```suggestion
    n ≤ G.chromaticNumber ↔ ∀ C : G.Coloring (Fin n), Function.Surjective C := by
```

File: Mathlib/Combinatorics/SimpleGraph/Coloring.lean [hunk 4]
```diff
@@ -392,6 +392,42 @@ theorem chromaticNumber_eq_card_of_forall_surj [Fintype α] (C : G.Coloring α)
...
+    · exfalso; exact C.valid had (h2 ▸ h1)
+    all_goals intro hne; simp only [Fin.mk.injEq, Fin.val_eq_val, hc] at hne
+    · exact hc b hne.symm
+    · exact C.valid had hne⟩
+    apply lt_of_le_of_lt (Colorable.chromaticNumber_le ⟨C'⟩)
+    rw [Nat.cast_lt]
+    exact Nat.lt_succ_self _
+
+/-- If G has a α-Coloring then χ(G) = ‖α‖ ↔ every α-Coloring of G is surjective -/
+theorem chromaticNumber_eq_card_iff_forall_surj [Fintype α] (C : G.Coloring α) :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Coloring.lean:419]:
```suggestion
theorem chromaticNumber_eq_card_iff_forall_surj :
```

File: Mathlib/Combinatorics/SimpleGraph/Coloring.lean [hunk 5]
```diff
@@ -392,6 +392,42 @@ theorem chromaticNumber_eq_card_of_forall_surj [Fintype α] (C : G.Coloring α)
...
+    G.chromaticNumber = Fintype.card α  ↔ (∀ C' : G.Coloring α, Function.Surjective C') := by
+  refine ⟨fun h C' ↦ ?_,fun h ↦ chromaticNumber_eq_card_of_forall_surj C h⟩
+  let e := Fintype.equivFin α
+  let D := G.recolorOfEquiv e C'
+  convert (e.symm.comp_surjective D).2 <| le_chromaticNumber_surj D h.symm.le
+  ext
+  apply_fun e
+  rw [Function.comp_apply, Equiv.apply_symm_apply]
+  rfl
+
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Coloring.lean:429]:
```suggestion
```

**Review by YaelDillies — COMMENTED** [review]
Your second theorem is just a combination of an inequality and the fact that `G.Colorable n` implies `n ≤ G.chromaticNumber`

**Comment by jt496** [comment]
Thanks for looking at this and for your suggestions! 

Can I just check I've understood what you mean by

> Your second theorem is just a combination of an inequality and the fact that `G.Colorable n` implies `n ≤ G.chromaticNumber`

By second theorem do you mean `theorem chromaticNumber_eq_card_iff_forall_surj`?

Also,  I don't see how `G.Colorable n` implies `n ≤ G.chromaticNumber`?

**Comment by jt496** [comment]
@YaelDillies I've tried to incorporate your suggestions.

I'm not sure the 2nd result is worth having, but I've left it in for now as :

`theorem le_chromaticNumber_iff_forall_surj (C : G.Coloring (Fin n)) :
    n ≤ G.chromaticNumber ↔ ∀ C : G.Coloring (Fin n), Function.Surjective C `

File: Mathlib/Combinatorics/SimpleGraph/Coloring.lean [hunk 6]
```diff
@@ -392,41 +392,26 @@ theorem chromaticNumber_eq_card_of_forall_surj [Fintype α] (C : G.Coloring α)
...
+  let D : G.Coloring {a : α // a ≠ i} := ⟨fun v ↦ ⟨C v, hi v⟩, by
+    intro _ _ hadj hne
+    exact C.valid hadj <| Subtype.mk.injEq _ _ _ _ ▸ hne⟩
+  classical
+  exact Nat.not_mem_of_lt_sInf (trans (Nat.pred_lt' <| card_pos_iff.2 ⟨i⟩) h)
+      ⟨G.recolorOfEquiv (equivOfCardEq <| card_fin _ ▸ (card_subtype_compl _)) D⟩
+
+variable {n : ℕ}
 
+theorem le_chromaticNumber_iff_forall_surj (C : G.Coloring (Fin n)) :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Coloring.lean:411]:
```suggestion
theorem le_chromaticNumber_iff_forall_surj :
```

**Comment by jt496** [comment]
Sorry, I'd misunderstood your earlier suggestion, I've now incorporated it.

**Review by YaelDillies — APPROVED** [review]
I've slightly cleaned up your new theorems and existing ones. I believe this is now good to go.

maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Comment by jt496** [comment]
Thanks!

**Pull request closed without merging**